hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/JmxServiceProviderImpl.java

```diff
@@ -0,0 +1,194 @@
+   * given metric.
+   *
+   * @param api         api.
+   * @param queryString query string with metric name and other filters.
+   * @return HttpURLConnection
+   */
+  @Override
+  public HttpURLConnection getMetricsResponse(String api, String queryString) {
+    return null;
+  }
+
+  @Override
+  public List<Metric> getMetricsInstant(String queryString) {
+    return Collections.emptyList();
+  }
+
+  /**
+   * Returns a list of {@link Metric} for the given instant query.
+   *
+   * @param queryString query string with metric name and other filters.
+   * @return List of Json map of metrics response.
+   */
+  @Override
+  public CompletableFuture<List<Map<String, Object>>> getMetricsAsync(String jmxEndpoint, String queryString) {
+    // Remove the trailing slash from endpoint url.
+    if (jmxEndpoint != null && jmxEndpoint.endsWith("/")) {
+      jmxEndpoint = jmxEndpoint.substring(0, jmxEndpoint.length() - 1);
+    }
+    return getMetrics(jmxEndpoint, queryString);
+  }
+
+  @Override
+  public void shutdown() {
+    try {
+      LOG.info("Shutting down async HTTP client...");
+      asyncHttpClient.close();
+    } catch (IOException e) {
+      LOG.error("Error shutting down async HTTP client", e);
+    }
+  }
+
+  /**
+   * Returns a list of {@link Metric} for the given api and query string.
+   *
+   * @param jmxEndpoint endpoint for the jmx server
+   * @param queryString query string with metric name and other filters.
+   * @return List of Json map of metrics response.
+   */
+  private CompletableFuture<List<Map<String, Object>>> getMetrics(String jmxEndpoint, String queryString) {
+    CompletableFuture<List<Map<String, Object>>> future = new CompletableFuture<>();
+
+    try {
+      String url = String.format(
+          "%s?%s=%s",
+          jmxEndpoint,
+          JMX_INSTANT_QUERY_API,
+          URLEncoder.encode(queryString, "UTF-8"));
+      HttpGet request = new HttpGet(url);
+
+      asyncHttpClient.execute(request, new FutureCallback<HttpResponse>() {
+        @Override
+        public void completed(HttpResponse response) {
+          try {
+            int statusCode = response.getStatusLine().getStatusCode();
+            if (statusCode >= 200 && statusCode < 300) {
+              String responseBody = EntityUtils.toString(response.getEntity());
+              Map<String, Object> jsonResponse = objectMapper.readValue(responseBody, Map.class);
+              List<Map<String, Object>> beans = (List<Map<String, Object>>) jsonResponse.get("beans");
+              future.complete(beans);
+            } else {
+              future.completeExceptionally(new IOException("HTTP error code: " + statusCode));
+            }
+          } catch (Exception e) {
+            future.completeExceptionally(e);
+          }
+        }
+
+        @Override
+        public void failed(Exception ex) {
+          LOG.debug("HTTP request failed for {}: {}", url, ex.getMessage());
```

Comment:
I feel, this may not be very bulk, so connection failures to datanodes should not be logged at DEBUG level. We can log at warn level atleast and log complete error trace instead of just message.